src/workerd/jsg/filesystem.h

```diff
@@ -0,0 +1,244 @@
+  // an exception will be thrown. The name must not contain any path separators.
+  // If the given file or directory has not been detached from its parent, an
+  // exception will be thrown. If this directory is read only, an exception will
+  // be thrown.
+  virtual kj::Own<File> transfer(Lock& js, kj::StringPtr name, kj::Own<File>&& file) = 0;
+  virtual kj::Own<Directory> transfer(Lock& js, kj::StringPtr name, kj::Own<Directory>&& file) = 0;
+
+  struct RemoveOptions {
+    // If true and the node is a directory, remove all entries in the directory.
+    bool recursive = false;
+  };
+
+  // Tries to remove the file or directory at the given path. If the node does
+  // not exist, true will be returned. If the node is a directory and the recursive
+  // option is not set, an exception will be thrown if the directory is not empty.
+  // If the directory is read only, an exception will be thrown.
+  // If the node is a file, it will be removed regardless of the recursive option
+  // if the directory is not read only.
+  // The path must be relative to the current directory.
+  // Note that this method will only remove the node from this directory. If the
+  // node as references in other directories, those will not be removed.
+  virtual bool remove(Lock& js, kj::PathPtr path, kj::Maybe<RemoveOptions> options = kj::none) = 0;
+
+  // Creates a new readable/writable in-memory directory. This directory will
+  // not be initially included in a directory. To add it to a directory, use the
+  // transfer method. If the directory is not transferred, it will be deleted
+  // when the handle is dropped.
+  static kj::Own<Directory> newWritable();
+
+  // Used to build a new read-only directory. All files and directories added
+  // may or may not be writable. The directory will not be initially included
+  // in a directory. To add it to a directory, use the transfer method.
+  class Builder {
+   public:
+    Builder() = default;
+    KJ_DISALLOW_COPY_AND_MOVE(Builder);
+
+    // Adds a file to the directory. The name must not contain any path separators.
+    // If the name already exists an exception is thrown.
+    void add(kj::StringPtr name, kj::Own<File>&& file);
+
+    // Adds a directory to the directory. The name must not contain any path separators.
+    // If the name already exists an exception is thrown.
+    void add(kj::StringPtr name, kj::Own<Directory>&& dir);
+
+    void add(kj::StringPtr name, kj::Own<Directory::Builder>&& dir);
+
+    void addPath(kj::PathPtr path, kj::Own<Directory>&& dir);
+    void addPath(kj::PathPtr path, kj::Own<File>&& file);
+
+    // Finalizes and returns the directory. The directory will not be writable.
+    // The directory will not be initially included in a directory. To add it
+    // to a directory, use the transfer method.
+    kj::Own<Directory> finish();
+
+    using Map = kj::HashMap<kj::String,
+        kj::OneOf<kj::Own<File>, kj::Own<Directory>, kj::Own<Directory::Builder>>>;
```

Comment:
`kj::OneOf< kj::Own<File>, kj::Own<Directory>>` could be more cleanly represented as `kj::Own<Node>`

Reply:
Same... I'm removing `Node` and will keep the `kj::OneOf`